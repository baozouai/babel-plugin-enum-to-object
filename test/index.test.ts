import { expect, it } from 'vitest'
import type { TransformOptions } from '@babel/core'
import { transformSync } from '@babel/core'
import plugin from '../src/babel-plugin-enum-to-object'

const defaultOptions: TransformOptions = {
  plugins: [plugin],
  generatorOpts: {
    jsescOption: {
      quotes: 'single',
    },
  },

}

const noReflectOptions: TransformOptions = {
  plugins: [[plugin, { reflect: false }]],
  generatorOpts: {
    jsescOption: {
      quotes: 'single',
    },
  },
}
const numberEnumInput = `export enum Direction {
  Up,
  Down,
  Left,
  Right,
}  
`
const stringEnumInput = `export enum Direction {
  Up = "Up",
  Down = "Down",
  Left = "Left",
  Right = "Right",
}`

const mixEnumInput = `export enum Status {
  PAID,
  UNPAID = '21',
  PART = 44,
  HALF
}
`

const calcEnumInput = `export enum MyEnum {
  A,
  B,
  C = 20,
  D,
  E = B * C * D * 10,
  F = E - 1,
  G = F + 1 + 2,
  H = G / 2,
  I = H % 2 + 4,
  J = E ** 3,
  K = E << 2,
  L = E >> 2,
  M = E >>> 1,
  N = E + 1 & 1,
  O = E | 2 + 1,
  P = E ^ 2,
}
`

const complexEnumInput = `
export enum Other {
  A = 4,
  B = 5,
  C = 6,
}
export enum MyEnum {
  'GP-2120TU' = Other.A,
  'GP-M32B' = Other.B,
  'GP-R222C' = Other.C,
}
`

const initByOtherIdentifiterEnumInput = `
const WW = 2
export enum MyEnum {
  'GP-2120TU' = WW,
  'GP-M32B' = 3,
  'GP-R222C',
}
`

it('Transforms NumericLiteral enum', () => {
  const { code } = transformSync(numberEnumInput, defaultOptions)!
  expect(code).toMatchSnapshot()
})

it('Transforms string enum', () => {
  const { code } = transformSync(stringEnumInput, defaultOptions)!
  expect(code).toMatchSnapshot()
})

it('Transforms string number mix', () => {
  const { code } = transformSync(mixEnumInput, defaultOptions)!
  expect(code).toMatchSnapshot()
})

it('Transforms calc enum', () => {
  const { code } = transformSync(calcEnumInput, defaultOptions)!
  expect(code).toMatchSnapshot()
})

it('Transforms complex enum', () => {
  const { code } = transformSync(complexEnumInput, defaultOptions)!
  expect(code).toMatchSnapshot()
})
it('Transforms initByOtherIdentifiter enum', () => {
  const { code } = transformSync(initByOtherIdentifiterEnumInput, defaultOptions)!
  expect(code).toMatchSnapshot()
})
it('Transforms NumericLiteral enum  with reflect false', () => {
  const { code } = transformSync(numberEnumInput, noReflectOptions)!
  expect(code).toMatchSnapshot()
})

it('Transforms string enum  with reflect false', () => {
  const { code } = transformSync(stringEnumInput, noReflectOptions)!
  expect(code).toMatchSnapshot()
})

it('Transforms string number mix  with reflect false', () => {
  const { code } = transformSync(mixEnumInput, noReflectOptions)!
  expect(code).toMatchSnapshot()
})

it('Transforms calc enum  with reflect false', () => {
  const { code } = transformSync(calcEnumInput, noReflectOptions)!
  expect(code).toMatchSnapshot()
})

it('Transforms complex enum  with reflect false', () => {
  const { code } = transformSync(complexEnumInput, noReflectOptions)!
  expect(code).toMatchSnapshot()
})

it('Transforms initByOtherIdentifiter enum false', () => {
  const { code } = transformSync(initByOtherIdentifiterEnumInput, noReflectOptions)!
  expect(code).toMatchSnapshot()
})