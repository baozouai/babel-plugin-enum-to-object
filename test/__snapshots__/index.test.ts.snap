// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transforms NumericLiteral enum  with reflect false 1`] = `
"export const Direction = {
  'Up': 0,
  'Down': 1,
  'Left': 2,
  'Right': 3
};"
`;

exports[`Transforms NumericLiteral enum 1`] = `
"export const Direction = {
  'Up': 0,
  '0': 'Up',
  'Down': 1,
  '1': 'Down',
  'Left': 2,
  '2': 'Left',
  'Right': 3,
  '3': 'Right'
};"
`;

exports[`Transforms calc enum  with reflect false 1`] = `
"export const MyEnum = {
  'A': 0,
  'B': 1,
  'C': 20,
  'D': 21,
  'E': 4200,
  'F': 4199,
  'G': 4202,
  'H': 2101,
  'I': 5,
  'J': 74088000000,
  'K': 16800,
  'L': 1050,
  'M': 2100,
  'N': 1,
  'O': 4203,
  'P': 4202
};"
`;

exports[`Transforms calc enum 1`] = `
"export const MyEnum = {
  'A': 0,
  '0': 'A',
  'B': 1,
  '1': 'N',
  'C': 20,
  '20': 'C',
  'D': 21,
  '21': 'D',
  'E': 4200,
  '4200': 'E',
  'F': 4199,
  '4199': 'F',
  'G': 4202,
  '4202': 'P',
  'H': 2101,
  '2101': 'H',
  'I': 5,
  '5': 'I',
  'J': 74088000000,
  '74088000000': 'J',
  'K': 16800,
  '16800': 'K',
  'L': 1050,
  '1050': 'L',
  'M': 2100,
  '2100': 'M',
  'N': 1,
  'O': 4203,
  '4203': 'O',
  'P': 4202
};"
`;

exports[`Transforms complex enum  with reflect false 1`] = `
"export const Other = {
  'A': 4,
  'B': 5,
  'C': 6
};
export const MyEnum = {
  'GP-2120TU': Other.A,
  'GP-M32B': Other.B,
  'GP-R222C': Other.C
};"
`;

exports[`Transforms complex enum 1`] = `
"export const Other = {
  'A': 4,
  '4': 'A',
  'B': 5,
  '5': 'B',
  'C': 6,
  '6': 'C'
};
export const MyEnum = {
  'GP-2120TU': Other.A,
  [Other.A]: 'GP-2120TU',
  'GP-M32B': Other.B,
  [Other.B]: 'GP-M32B',
  'GP-R222C': Other.C,
  [Other.C]: 'GP-R222C'
};"
`;

exports[`Transforms initByOtherIdentifiter enum 1`] = `
"const WW = 2;
export const MyEnum = {
  'GP-2120TU': WW,
  [WW]: 'GP-2120TU',
  'GP-M32B': 3,
  '3': 'GP-M32B',
  'GP-R222C': 4,
  '4': 'GP-R222C'
};"
`;

exports[`Transforms initByOtherIdentifiter enum false 1`] = `
"const WW = 2;
export const MyEnum = {
  'GP-2120TU': WW,
  'GP-M32B': 3,
  'GP-R222C': 4
};"
`;

exports[`Transforms string enum  with reflect false 1`] = `
"export const Direction = {
  'Up': 'Up',
  'Down': 'Down',
  'Left': 'Left',
  'Right': 'Right'
};"
`;

exports[`Transforms string enum 1`] = `
"export const Direction = {
  'Up': 'Up',
  'Down': 'Down',
  'Left': 'Left',
  'Right': 'Right'
};"
`;

exports[`Transforms string number mix  with reflect false 1`] = `
"export const Status = {
  'PAID': 0,
  'UNPAID': '21',
  'PART': 44,
  'HALF': 45
};"
`;

exports[`Transforms string number mix 1`] = `
"export const Status = {
  'PAID': 0,
  '0': 'PAID',
  'UNPAID': '21',
  '21': 'UNPAID',
  'PART': 44,
  '44': 'PART',
  'HALF': 45,
  '45': 'HALF'
};"
`;
